File: algorithms/2_medium/D/00641_design_circular_deque/deque.c
typedef struct {
    size_t len;
    size_t head;
    size_t cap;
    int buff[];
} MyCircularDeque;

bool myCircularDequeIsEmpty(MyCircularDeque* obj);
bool myCircularDequeIsFull(MyCircularDeque* obj);

MyCircularDeque* myCircularDequeCreate(int k)
{
    assert(k > 0);
    MyCircularDeque* res = malloc(sizeof(*res) + (size_t)k * sizeof(int));
    assert(res != NULL);
    res->len = 0;
    res->cap = (size_t)k;
    res->head = 0;
    return res;
}

bool myCircularDequeInsertFront(MyCircularDeque* obj, int val)
{
    if (myCircularDequeIsFull(obj)) {
        return false;
    }
    const size_t head = obj->head;
    const size_t cap = obj->cap;
    const size_t idx = (head + cap - 1) % cap;
    obj->buff[idx] = val;
    obj->head = idx;
    obj->len += 1;
    return true;
}

bool myCircularDequeInsertLast(MyCircularDeque* obj, int val)
{
    if (myCircularDequeIsFull(obj)) {
        return false;
    }
    const size_t head = obj->head;
    const size_t cap = obj->cap;
    const size_t len = obj->len;
    const size_t idx = (head + len) % cap;
    obj->buff[idx] = val;
    obj->len += 1;
    return true;
}

bool myCircularDequeDeleteFront(MyCircularDeque* obj)
{
    if (myCircularDequeIsEmpty(obj)) {
        return false;
    }
    obj->head = (obj->head + 1) % obj->cap;
    obj->len -= 1;
    return true;
}

bool myCircularDequeDeleteLast(MyCircularDeque* obj)
{
    if (myCircularDequeIsEmpty(obj)) {
        return false;
    }
    obj->len -= 1;
    return true;
}

int myCircularDequeGetFront(MyCircularDeque* obj)
{
    if (myCircularDequeIsEmpty(obj)) {
        return -1;
    }
    return obj->buff[obj->head];
}

int myCircularDequeGetRear(MyCircularDeque* obj) {
    if (myCircularDequeIsEmpty(obj)) {
        return -1;
    }
    return obj->buff[(obj->head + obj->len + obj->cap - 1) % obj->cap];
}

bool myCircularDequeIsEmpty(MyCircularDeque* obj)
{
    return obj->len == 0;
}

bool myCircularDequeIsFull(MyCircularDeque* obj)
{
    return obj->len == obj->cap;
}

void myCircularDequeFree(MyCircularDeque* obj)
{
    free(obj);
}

/**
 * Your MyCircularDeque struct will be instantiated and called as such:
 * MyCircularDeque* obj = myCircularDequeCreate(k);
 * bool param_1 = myCircularDequeInsertFront(obj, value);
 
 * bool param_2 = myCircularDequeInsertLast(obj, value);
 
 * bool param_3 = myCircularDequeDeleteFront(obj);
 
 * bool param_4 = myCircularDequeDeleteLast(obj);
 
 * int param_5 = myCircularDequeGetFront(obj);
 
 * int param_6 = myCircularDequeGetRear(obj);
 
 * bool param_7 = myCircularDequeIsEmpty(obj);
 
 * bool param_8 = myCircularDequeIsFull(obj);
 
 * myCircularDequeFree(obj);
*/

File: algorithms/2_medium/D/00641_design_circular_deque/deque.rs
struct MyCircularDeque {
    buff: Vec<i32>,
    len: usize,
    head: usize,
    cap: usize,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MyCircularDeque {

    fn new(k: i32) -> Self {
        assert!(k > 0);
        let mut buff = Vec::with_capacity(k as usize);
        unsafe { buff.set_len(k as usize); }
        Self {
            buff: buff,
            len: 0,
            head: 0, 
            cap: k as usize,
        }
    }

    fn insert_front(&mut self, val: i32) -> bool {
        if self.is_full() {
            return false;
        }
        let (head, cap) = (self.head, self.cap);
        let idx = (head + cap - 1) % cap;
        self.buff[idx] = val;
        self.head = idx;
        self.len += 1;
        true
    }

    fn insert_last(&mut self, val: i32) -> bool {
        if self.is_full() {
            return false;
        }
        let (head, len, cap) = (self.head, self.len, self.cap);
        let idx = (head + len) % cap;
        self.buff[idx] = val;
        self.len += 1;
        true
    }

    fn delete_front(&mut self) -> bool {
        if self.is_empty() {
            return false;
        }
        self.head = (self.head + 1) % self.cap;
        self.len -= 1;
        true
    }

    fn delete_last(&mut self) -> bool {
        if self.is_empty() {
            return false;
        }
        self.len -= 1;
        true
    }

    fn get_front(&self) -> i32 {
        if self.is_empty() {
            return -1;
        }
        self.buff[self.head]
    }

    fn get_rear(&self) -> i32 {
        if self.is_empty() {
            return -1;
        }
        self.buff[(self.head + self.len + self.cap - 1) % self.cap]
    }

    fn is_empty(&self) -> bool {
        self.len == 0
    }

    fn is_full(&self) -> bool {
        self.cap == self.len
    }
}

/**
 * Your MyCircularDeque object will be instantiated and called as such:
 * let obj = MyCircularDeque::new(k);
 * let ret_1: bool = obj.insert_front(value);
 * let ret_2: bool = obj.insert_last(value);
 * let ret_3: bool = obj.delete_front();
 * let ret_4: bool = obj.delete_last();
 * let ret_5: i32 = obj.get_front();
 * let ret_6: i32 = obj.get_rear();
 * let ret_7: bool = obj.is_empty();
 * let ret_8: bool = obj.is_full();
 */
